File: src/ai/openai/codex/frontend/client/ClientConnection.cpp
/*
 * SPDX-License-Identifier: LGPL-3.0-or-later OR MIT
 */

#include "ai/openai/codex/frontend/client/ClientConnection.h"

#include "ai/openai/codex/frontend/CodexBridge.h"

#include <utility>

namespace ai::openai::codex::frontend::client {

    ClientConnection::ClientConnection(frontend::CodexBridge& sdk, ClientConnectionCallbacks callbacks)
        : sdk_(sdk)
        , callbacks_(std::move(callbacks)) {
        sdk_.setSender([this](nlohmann::json message) { return send(std::move(message)); });
    }

    ClientConnection::~ClientConnection() {
        shutdown();
        sdk_.setSender({});
    }

    bool ClientConnection::attach(TransportEndpoint& endpoint) noexcept {
        if (shuttingDown_ || (endpoint_ != nullptr && endpoint_ != &endpoint)) {
            return false;
        }
        endpoint_ = &endpoint;
        failureReported_ = false;
        return true;
    }

    void ClientConnection::connected(TransportEndpoint& endpoint) noexcept {
        if (endpoint_ != &endpoint || shuttingDown_ || online_) {
            return;
        }
        online_ = true;
        try {
            if (callbacks_.onConnected) {
                callbacks_.onConnected();
            }
        } catch (...) {
        }
    }

    void ClientConnection::receive(TransportEndpoint& endpoint, nlohmann::json message) noexcept {
        if (endpoint_ != &endpoint || !online_ || shuttingDown_) {
            return;
        }
        try {
            if (!sdk_.receive(message)) {
                failed(endpoint, "bridge client received an invalid envelope");
            }
        } catch (...) {
            failed(endpoint, "bridge client failed while dispatching an envelope");
        }
    }

    void ClientConnection::failed(TransportEndpoint& endpoint, std::string reason) noexcept {
        if (endpoint_ != &endpoint) {
            return;
        }
        reportFailure(reason);
        endpoint.close(reason);
    }

    void ClientConnection::detach(TransportEndpoint& endpoint, std::string reason) noexcept {
        if (endpoint_ != &endpoint) {
            return;
        }
        endpoint_ = nullptr;
        const bool wasOnline = std::exchange(online_, false);
        failureReported_ = false;
        sdk_.transportDisconnected(shuttingDown_ ? "bridge client shutdown" : reason);
        if (wasOnline) {
            try {
                if (callbacks_.onDisconnected) {
                    callbacks_.onDisconnected();
                }
            } catch (...) {
            }
        }
    }

    void ClientConnection::disconnect(std::string_view reason) noexcept {
        if (!shuttingDown_ && endpoint_ != nullptr) {
            endpoint_->close(reason);
        }
    }

    void ClientConnection::shutdown() noexcept {
        if (shuttingDown_) {
            return;
        }
        shuttingDown_ = true;
        if (endpoint_ != nullptr) {
            endpoint_->close("bridge client shutdown");
        } else {
            sdk_.transportDisconnected("bridge client shutdown");
        }
    }

    bool ClientConnection::online() const noexcept {
        return online_;
    }

    bool ClientConnection::attached() const noexcept {
        return endpoint_ != nullptr;
    }

    bool ClientConnection::send(nlohmann::json message) noexcept {
        if (!online_ || shuttingDown_ || endpoint_ == nullptr) {
            return false;
        }
        try {
            return endpoint_->send(message);
        } catch (...) {
            return false;
        }
    }

    void ClientConnection::reportFailure(std::string reason) noexcept {
        if (failureReported_) {
            return;
        }
        failureReported_ = true;
        try {
            if (callbacks_.onFailure) {
                callbacks_.onFailure(std::move(reason));
            }
        } catch (...) {
        }
    }

} // namespace ai::openai::codex::frontend::client

File: packages/codex-frontend/src/ClientConnection.ts
import type {CodexBridgeClient} from "./CodexBridgeClient.js";
import type {BridgeMessage} from "./protocol/envelope.js";

export interface TransportEndpoint {
    send(message: BridgeMessage): boolean;
    close(reason: string): void;
}

export interface ClientConnectionCallbacks {
    readonly onConnected?: () => void;
    readonly onDisconnected?: () => void;
    readonly onFailure?: (reason: string) => void;
}

export class ClientConnection {
    private endpoint: TransportEndpoint | undefined;
    private isOnline = false;
    private isShuttingDown = false;
    private failureReported = false;

    public constructor(
        private readonly sdk: CodexBridgeClient,
        private readonly callbacks: ClientConnectionCallbacks = {},
    ) {
        this.sdk.setSender((message) => this.send(message));
    }

    public attach(endpoint: TransportEndpoint): boolean {
        if (
            this.isShuttingDown
            || (this.endpoint !== undefined && this.endpoint !== endpoint)
        ) {
            return false;
        }
        this.endpoint = endpoint;
        this.failureReported = false;
        return true;
    }

    public connected(endpoint: TransportEndpoint): void {
        if (this.endpoint !== endpoint || this.isShuttingDown || this.isOnline) return;
        this.isOnline = true;
        this.invokeContained(this.callbacks.onConnected);
    }

    public receive(endpoint: TransportEndpoint, message: unknown): void {
        if (this.endpoint !== endpoint || !this.isOnline || this.isShuttingDown) return;
        try {
            if (!this.sdk.receive(message)) {
                this.failed(endpoint, "bridge client received an invalid envelope");
            }
        } catch {
            this.failed(endpoint, "bridge client failed while dispatching an envelope");
        }
    }

    public failed(endpoint: TransportEndpoint, reason: string): void {
        if (this.endpoint !== endpoint) return;
        this.reportFailure(reason);
        try {
            endpoint.close(reason);
        } catch {
            // Transport close failures cannot escape lifecycle dispatch.
        }
    }

    public detach(endpoint: TransportEndpoint, reason: string): void {
        if (this.endpoint !== endpoint) return;
        this.endpoint = undefined;
        const wasOnline = this.isOnline;
        this.isOnline = false;
        this.failureReported = false;
        this.sdk.transportDisconnected(
            this.isShuttingDown ? "bridge client shutdown" : reason,
        );
        if (wasOnline) this.invokeContained(this.callbacks.onDisconnected);
    }

    public disconnect(reason: string): void {
        if (!this.isShuttingDown && this.endpoint !== undefined) {
            try {
                this.endpoint.close(reason);
            } catch {
                // Explicit disconnect remains noexcept at the application boundary.
            }
        }
    }

    public shutdown(): void {
        if (this.isShuttingDown) return;
        this.isShuttingDown = true;
        if (this.endpoint !== undefined) {
            try {
                this.endpoint.close("bridge client shutdown");
            } catch {
                // Shutdown remains contained even when the transport misbehaves.
            }
        } else {
            this.sdk.transportDisconnected("bridge client shutdown");
        }
    }

    public dispose(): void {
        this.shutdown();
        this.sdk.setSender(undefined);
    }

    public get attached(): boolean {
        return this.endpoint !== undefined;
    }

    public get online(): boolean {
        return this.isOnline;
    }

    private send(message: BridgeMessage): boolean {
        if (!this.isOnline || this.isShuttingDown || this.endpoint === undefined) {
            return false;
        }
        try {
            return this.endpoint.send(message);
        } catch {
            return false;
        }
    }

    private reportFailure(reason: string): void {
        if (this.failureReported) return;
        this.failureReported = true;
        this.invokeContained(this.callbacks.onFailure, reason);
    }

    private invokeContained<Arguments extends readonly unknown[]>(
        callback: ((...args: Arguments) => void) | undefined,
        ...args: Arguments
    ): void {
        if (callback === undefined) return;
        try {
            callback(...args);
        } catch {
            // Application lifecycle callbacks are isolated from the transport.
        }
    }
}
